crates/core/tedge_actors/src/channels.rs

```diff
@@ -0,0 +1,226 @@
+use crate::{ChannelError, Message};
+use async_trait::async_trait;
+use futures::channel::mpsc;
+use futures::{SinkExt, StreamExt};
+use std::fmt::{Debug, Formatter};
+
+/// Create a new mailbox with its address
+///
+/// Such a mailbox is used by an actor to receive all its messages.
+/// Clones of the address are given to the sending peers.
+pub fn new_mailbox<M>(bound: usize) -> (Mailbox<M>, Address<M>) {
+    let (sender, receiver) = mpsc::channel(bound);
+    let address = Address { sender };
+    let mailbox = Mailbox { receiver };
+    (mailbox, address)
+}
+
+/// A mailbox that gather *all* the messages sent to an actor
+pub struct Mailbox<M> {
+    receiver: mpsc::Receiver<M>,
+}
+
+impl<M> Mailbox<M> {
+    /// Pop from the mailbox the message with the highest priority
+    ///
+    /// Await till a messages is available.
+    /// Return `None` when all the senders to this mailbox have been dropped and all the messages consumed.
+    pub async fn next(&mut self) -> Option<M> {
+        self.receiver.next().await
+    }
+
+    /// Collect all the messages of the mailbox into a vector
+    ///
+    /// Mostly useful for testing.
+    /// Note that this will block until there is no more senders,
+    /// .i.e. the mailbox address and all its clones have been dropped.
+    pub async fn collect(mut self) -> Vec<M> {
+        let mut messages = vec![];
+        while let Some(message) = self.next().await {
+            messages.push(message);
+        }
+        messages
+    }
+}
+
+/// The address of an actor
+pub struct Address<M> {
+    sender: mpsc::Sender<M>,
+}
+
+// The derive macro incorrectly requires M to be Clone
+impl<M> Clone for Address<M> {
+    fn clone(&self) -> Self {
+        Address {
+            sender: self.sender.clone(),
+        }
+    }
+}
+
+impl<M: Message> Address<M> {
+    /// Build a clone of this address to used as a recipient of sub-messages,
+    /// i.e. messages that can be cast into those expected by the mailbox.
+    pub fn as_recipient<N: Message + Into<M>>(&self) -> Recipient<N> {
+        self.clone().into()
+    }
+}
+
+/// A recipient for messages of type `M`
+///
+/// Actors don't access directly the addresses of their peers,
+/// but use intermediate recipients that adapt the messages when sent.
+pub type Recipient<M> = Box<dyn Sender<M>>;
```

Comment:
As I see this `Recipient` type more and more, it feels like this type indirection is making the code harder to follow. Using the `mpsc::Sender` and `mpsc::Receiver` abstractions directly feel more natural to me. I'm even questioning the relevance of these "pluggable message adaption". When the `ConfigManagerActor` actor needs to send an `HttpRequest` to the `HttpActor`, why not expect the `ConfigManagerActor` itself to create the `HttpRequest` message and push it to the `HttpActor`'s mailbox via a `Sender` of that actor?

Reply:
> When the `ConfigManagerActor` actor needs to send an `HttpRequest` to the `HttpActor`, why not expect the `ConfigManagerActor` itself to create the `HttpRequest` message and push it to the `HttpActor`'s mailbox via a `Sender` of that actor?

This is exactly what is done. At least from the `ConfigManagerActor` point of view, i.e. the client point of view.

The role of a *`dyn`* adapter is on the recipient side. So, the `HttpActor` can multiplex requests from several clients and a complex actor as the c8y mapper can handle different types of requests coming from independent sources.

> As I see this Recipient type more and more, it feels like this type indirection is making the code harder to follow.
> Using the `mpsc::Sender` and `mpsc::Receiver` abstractions directly feel more natural to me.

I note your concerns. Something needs to be improved / simplified. Either the design or the explanations.

What you propose here is aligned with your previous remark about `add_client` returning a `mpsc::Receiver`.
This leads to a model with one `mpsc::Receiver` and one `mpsc::Sender` per peer. This can be explored.

On my side, I also consider to simplify this recipient type but still using dyn object and not concrete type as `mpsc::Sender`.

```
pub type Recipient<M> = Box<dyn futures::sink::Sink<M>>;
```